include/ginkgo/core/config/registry.hpp

```diff
@@ -0,0 +1,220 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_REGISTRY_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_REGISTRY_HPP_
+
+
+#include <complex>
+#include <functional>
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils_helper.hpp>
+#include <ginkgo/core/config/property_tree.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+namespace gko {
+namespace config {
+
+
+class registry;
+
+/**
+ * type_descriptor gives the initial default type for common type in ginkgo such
+ * as ValueType and IndexType. If the factory config does not specify these
+ * type, the configuration will use this as the default.
+ */
+class type_descriptor {
+public:
+    /**
+     * type_descriptor constructor. The correctness is checked in
+     * factory configuration. There is free function `make_type_descriptor` to
+     * create the object by template.
+     *
+     * @param value_typestr  the value type string. "void" means no default.
+     * @param index_typestr  the index type string. "void" means no default.
+     *
+     * @note there is no way to call the constructor with explicit template, so
+     * we create another free function to handle it.
+     */
+    explicit type_descriptor(std::string value_typestr = "void",
+                             std::string index_typestr = "void");
+
+    /**
+     * Get the value type string.
+     */
+    const std::string& get_value_typestr() const;
+
+    /**
+     * Get the index type string
+     */
+    const std::string& get_index_typestr() const;
+
+private:
+    std::string value_typestr_;
+    std::string index_typestr_;
+};
+
+
+template <typename ValueType = void, typename IndexType = void>
+type_descriptor make_type_descriptor();
+
+
+using linop_map = std::unordered_map<std::string, std::shared_ptr<LinOp>>;
+using linopfactory_map =
+    std::unordered_map<std::string, std::shared_ptr<LinOpFactory>>;
+using criterionfactory_map =
+    std::unordered_map<std::string, std::shared_ptr<stop::CriterionFactory>>;
+using buildfromconfig_map =
```

Comment:
This name is not great, maybe
```suggestion
using configuration_map =
```

Reply:
Also, maybe move this to another file. I'm not sure this is that closely related to the registry.